### PR TITLE
Feature/custom pdc api types

### PR DESCRIPTION
**Pull request by Sefiraat**
## Description
Allows for the use of custom PersistentDataType classes within the PersistentDataAPI

## Proposed changes
A new setter, has and getter(s) accepting a custom PersistentDataType and generic.

Also removed a duplicated @param

## Related Issues (if applicable)
N/A

**Note**: Tested by using the methods in a different plugin before copying here.

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: dough-data/src/main/java/io/github/bakedlibs/dough/data/persistent/PersistentDataAPI.java [hunk 1]
```diff
@@ -106,6 +106,24 @@ public static void setUUID(@Nonnull PersistentDataHolder holder, @Nonnull Namesp
...
+     * @param holder
+     *            The {@link PersistentDataHolder} to add the data to
+     * @param key
+     *            The key of the data to set
+     * @param type
+     *            The {@link PersistentDataType} to be used.
+     * @param obj
+     *            The object to put in the container
+     */
+
```
Comment (WalshyDev) [inline dough-data/src/main/java/io/github/bakedlibs/dough/data/persistent/PersistentDataAPI.java:121]: Comments should be attached to methods
```suggestion
```

File: dough-data/src/main/java/io/github/bakedlibs/dough/data/persistent/PersistentDataAPI.java [hunk 2]
```diff
@@ -316,6 +334,22 @@ public static boolean hasUUID(@Nonnull PersistentDataHolder holder, @Nonnull Nam
...
+     * Checks if the specified {@link PersistentDataHolder} has a {@link PersistentDataContainer} with the specified
+     * key.
+     *
+     * @param holder
+     *            The {@link PersistentDataHolder} to check
+     * @param key
+     *            The key to check for
+     * @return {@code true} if the holder has a {@link PersistentDataContainer} with the specified key.
+     */
+
```
Comment (WalshyDev) [inline dough-data/src/main/java/io/github/bakedlibs/dough/data/persistent/PersistentDataAPI.java:347]:
```suggestion
```

**Review by WalshyDev — CHANGES_REQUESTED** [review]

**Review by WalshyDev — DISMISSED** [review]
LGTM

not sure this is actually smaller than the Spigot API but at least it would create consistency

**Review by TheBusyBiscuit — COMMENTED** [review]
Do we even need to name this "custom" here?
I think get, set etc... itself is sufficient since you don't specify a "custom" type here, you just specify... *the* type.

**Review by WalshyDev — APPROVED** [review]
LGTM

**Review by TheBusyBiscuit — APPROVED** [review]

**Pull request merged**